File: src/ai/flows/understand-query.ts
// This is an AI-powered function to determine whether a user query is a request for data or a general question.
// - understandQuery - A function that handles the plant diagnosis process.
// - UnderstandQueryInput - The input type for the understandQuery function.
// - UnderstandQueryOutput - The return type for the understandQuery function.

'use server';

import {ai} from '@/ai/genkit';
import {z} from 'genkit';

const UnderstandQueryInputSchema = z.object({
  query: z.string().describe('The user query to understand.'),
});
export type UnderstandQueryInput = z.infer<typeof UnderstandQueryInputSchema>;

const UnderstandQueryOutputSchema = z.object({
  isDataRequest: z
    .boolean()
    .describe(
      'True if the query is a request for data from a database, false if it is a general question.'
    ),
  databaseType: z
    .string()
    .optional()
    .describe(
      'The type of database to query, e.g., MySQL, PostgreSQL, BigQuery. Only applicable if isDataRequest is true.'
    ),
  queryDetails: z
    .string()
    .optional()
    .describe(
      'Details about the data being requested. Only applicable if isDataRequest is true.'
    ),
});
export type UnderstandQueryOutput = z.infer<typeof UnderstandQueryOutputSchema>;

export async function understandQuery(input: UnderstandQueryInput): Promise<UnderstandQueryOutput> {
  return understandQueryFlow(input);
}

const prompt = ai.definePrompt({
  name: 'understandQueryPrompt',
  input: {schema: UnderstandQueryInputSchema},
  output: {schema: UnderstandQueryOutputSchema},
  prompt: `You need to determine if the following query is a request for data from a database, or a general question.

Query: {{{query}}}

Respond with a JSON object. The isDataRequest field should be true if the query is a request for data and false otherwise. If isDataRequest is true, populate the databaseType and queryDetails fields, otherwise they should be omitted.  The databaseType field should be the type of database to query, e.g., MySQL, PostgreSQL, BigQuery. The queryDetails field should be details about the data being requested.

Example 1:
Query: What is the capital of France?
{
  "isDataRequest": false
}

Example 2:
Query: Show me the average sales for the last month from the sales table.
{
  "isDataRequest": true,
  "databaseType": "Unknown",
  "queryDetails": "Average sales for the last month"
}

Example 3:
Query: list all customers from the customers table
{
  "isDataRequest": true,
  "databaseType": "Unknown",
  "queryDetails": "all customers"
}

Output:
`,
});

const understandQueryFlow = ai.defineFlow(
  {
    name: 'understandQueryFlow',
    inputSchema: UnderstandQueryInputSchema,
    outputSchema: UnderstandQueryOutputSchema,
  },
  async input => {
    const {output} = await prompt(input);
    return output!;
  }
);

File: src/ai/understand_query.py
import json
from typing import Dict, Any, Optional

# Assume 'llm' is an initialized large language model instance
# For example, using a hypothetical 'some_llm_library':
# from some_llm_library import LLM
# llm = LLM(model_name="your-preferred-model")

def understand_query(query: str) -> Dict[str, Any]:
    """
    Determines if a user query is a data request or a general question using an LLM.

    Args:
        query: The user's query string.

    Returns:
        A dictionary with 'isDataRequest' boolean and optional 'databaseType' and
        'queryDetails' strings.
    """
    prompt = f"""You need to determine if the following query is a request for data from a database, or a general question.

Query: {query}

Respond with a JSON object. The isDataRequest field should be true if the query is a request for data and false otherwise. If isDataRequest is true, populate the databaseType and queryDetails fields, otherwise they should be omitted.  The databaseType field should be the type of database to query, e.g., MySQL, PostgreSQL, BigQuery. The queryDetails field should be details about the data being requested.

Example 1:
Query: What is the capital of France?
{{
  "isDataRequest": false
}}

Example 2:
Query: Show me the average sales for the last month from the sales table.
{{
  "isDataRequest": true,
  "databaseType": "Unknown",
  "queryDetails": "Average sales for the last month"
}}

Example 3:
Query: list all customers from the customers table
{{
  "isDataRequest": true,
  "databaseType": "Unknown",
  "queryDetails": "all customers"
}}

Output:
"""

    # Replace this with your actual LLM call
    # For demonstration, we'll simulate an LLM response
    # llm_response = llm.generate(prompt)
    # response_text = llm_response.text # Or similar method to get string output

    # Simulated LLM response for testing
    if "sales" in query.lower() or "customers" in query.lower():
         simulated_response_text = """
{
  "isDataRequest": true,
  "databaseType": "Unknown",
  "queryDetails": "Data related to sales or customers"
}
"""
    else:
        simulated_response_text = """
{
  "isDataRequest": false
}
"""


    try:
        response_json = json.loads(simulated_response_text)
        return response_json
    except json.JSONDecodeError:
        # Handle cases where the LLM doesn't return valid JSON
        print(f"Warning: LLM returned invalid JSON: {simulated_response_text}")
        return {"isDataRequest": False}

if __name__ == '__main__':
    # Example Usage:
    query1 = "What is the weather like today?"
    result1 = understand_query(query1)
    print(f"Query: {query1}")
    print(f"Result: {result1}\n")

    query2 = "Show me the total sales for the last quarter from the orders table."
    result2 = understand_query(query2)
    print(f"Query: {query2}")
    print(f"Result: {result2}\n")

    query3 = "List all users in the system."
    result3 = understand_query(query3)
    print(f"Query: {query3}")
    print(f"Result: {result3}\n")